### ISSUE TITLE
[CMake] "No rule to make target" from 2.0.4 onwards

### ISSUE THREAD

**Issue by Razakhel** [emscripten-core/emsdk#676]
Hey there :wave: 

I've been trying to solve this issue for quite a long time, and as I'm still not successful, I figured I'd rather come here ask for advice and/or report a regression.

I've been using Emscripten for a while, and since last september [my builds all failed](https://travis-ci.org/github/Razakhel/RaZ/jobs/728403749#L4856) for still unknown reasons.

After having tracked down the faulty version, I can say for sure that this comes from the 2.0.4. Using the 2.0.3 [works just fine](https://github.com/Razakhel/RaZ/runs/1623478993).

Using CMake to build a project in C++, I'm developing a library on which example executables link themselves. Problem is, since Emscripten 2.0.4, I constantly get a message saying `No rule to make target <library>.bc, needed by <example>.html`.

The .bc extension is manually defined and seems to be irrelevant anyway, since I still get this issue whether I build an .a, .o, .lib or whatever. The problem "simply" seems to be that no library file is generated whatsoever. I've of course been trying to fix this locally, be it under Windows or Linux, but never has a lib file been produced ever since this version.

I'm using [`emcmake cmake`](https://github.com/Razakhel/RaZ/blob/bfd7e5d3426b2963b7957793e606af8dd6ea2a3a/.github/workflows/RaZ.yml#L199-L204) to generate the project. The specific CMake settings are [available here](https://github.com/Razakhel/RaZ/blob/bfd7e5d3426b2963b7957793e606af8dd6ea2a3a/CMakeLists.txt#L32-L35), while the library settings are [defined here](https://github.com/Razakhel/RaZ/blob/bfd7e5d3426b2963b7957793e606af8dd6ea2a3a/CMakeLists.txt#L193-L213). As for the demos, their extension is [set here](https://github.com/Razakhel/RaZ/blob/bfd7e5d3426b2963b7957793e606af8dd6ea2a3a/examples/CMakeLists.txt#L27-L31) and, if relevant, their assets packaging is [managed here](https://github.com/Razakhel/RaZ/blob/bfd7e5d3426b2963b7957793e606af8dd6ea2a3a/examples/EmscriptenAssets.cmake).

As a minimal example (here under Windows):

```cpp
// lib.hpp
void foo();

// lib.cpp
#include "lib.hpp"
void foo() {}
```
```
> em++ lib.cpp -o lib.a
> dir
29/12/2020  22:56    <DIR>          .
29/12/2020  22:56    <DIR>          ..
29/12/2020  22:55                34 lib.cpp
29/12/2020  22:53                12 lib.hpp
29/12/2020  22:56           110 142 lib.js
29/12/2020  22:56               674 lib.wasm
```

Are there any pointers anyone can give me to try to solve this issue? How can I force Emscripten to output a library to be linked to?

**Comment by sbc100** [comment]
;tldr; I think you need to remove your CMAKE_*_CREATE_STATIC_LIBRARY rules.

When you run a command of the form `em++ <source> -o <output>`, *without *`-c`, you are asking emscripten to produce an executable, and the format defaults to JavaScript.   If you run `em++ lib.cpp -o lib.a` emscripten will write a file called `lib.a` but it will contain a javascript program.    If you want to be a library archive you need to use `emar`.   If you want to build an object file you need to use `-c`.

This behavior did change at some point, in order to be more compatible with existing compilers list clang and gcc.    In that past we had special behavior for `-o <somehting>.bc` even in the absence of `-c`, but that is not longer present.

From the changelog is looks like this happened in 2.0.3:

```
- The default output format is now executable JavaScript.  Previously we would   
  default to output objecting files unless, for example, the output name ended   
  in `.js`.  This is contrary to behaviour of clang and gcc.  Now emscripten        
  will always produce and executable unless the `-c`, `-r` or `-shared` flags       
  are given.  This is true even when the name of the output file ends in `.o`.   
  e.g, `emcc foo.c -o foo.o` will produce a JavaScript file called `foo.o`.         
  This might surprise some users (although it matches the behavior of existing   
  toolchains) so we now produce a warning in this case.    
```

**Comment by sbc100** [comment]
BTW, running `em++ lib.cpp -o lib.a` should always produce a file called `lib.a` .. so I'm not sure why it isn't for you.. its certainly working for me right now.

**Comment by Razakhel** [comment]
TL;DR: I finally fixed this, the CMAKE_AR had to be left alone as well :ok_hand: 

> ;tldr; I think you need to remove your CMAKE_*_CREATE_STATIC_LIBRARY rules.

Doing so gives the following error: `emcc: error: qc: No such file or directory ("qc" was expected to be an input file, based on the commandline arguments provided)`. This was simply due to the fact that the called script wasn't `emar` but `emcc`.

> BTW, running em++ lib.cpp -o lib.a should always produce a file called lib.a .. so I'm not sure why it isn't for you.. its certainly working for me right now.

I guess I have an explanation for this. Locally, I still had the 2.0.4, which wouldn't give me a library file whatever I tried. I've just updated Emscripten to 2.0.11 and I now get one as expected. There may well have been an actual regression at that time...

---

Alright, I got it! The solution was simply to remove the CMAKE_*_CREATE_STATIC_LIBRARY rules ***and*** the CMAKE_AR, which was overriding `emar`. Doing so produces my library on which the executables link themselves instantly. Thanks a lot for your help!

I'm still confused about something though:

> In that past we had special behavior for -o <somehting>.bc even in the absence of -c, but that is not longer present.

Even without specifying `-c`, I still get a valid (and by that I mean an actual library, not a JS file with a different extension) .bc file with the 2.0.11. Should I always specify it for a library anyway?

As a last question, is there any difference outputting a .bc instead of leaving it to the default (outputs a .a in my current case)?

Thanks again for your help and Emscripten as a whole, it rocks (as long as I understand what to do) :+1:

**Comment by sbc100** [comment]
`.bc` files are kind of like combined object files.   We used to use them a lot in emscripten but we don't recommend that anymore.  These combined object files are produced by linking one or more object together (internally using `llvm-link` or `wasm-ld -r`).   Unless you are really sure of what you what you should avoid these.

Library files (ar arhives, which normally end in `.a`) are more standard and are more like zip files that contain many object files.  They also have some subtly different link semantics, for example in an library file only object that are actally needs are linked and unneeded object are completely ignored.

**Comment by Razakhel** [comment]
Alright, I'll leave the default extension then. Thanks a lot!
